Add WooCommerce > Subscriptions admin menu item and empty state offer screen

james-allan opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

**Internal project:** pdjTHR-Vi-p2

This PR adds a Subscriptions item to the WooCommerce admin menu for eligible stores * . This admin screen is part of an experiment to offer eligible merchants WooCommerce Payments built in Subscriptions functionality. 

\* eligible stores are stores without any subscriptions functionality plugin installed, are older than 6 months and have at least 1 sale in the last 30 days. 

<p align="center">
<img width="1161" alt="Screen Shot 2022-05-11 at 2 45 09 pm" src="https://user-images.githubusercontent.com/8490476/167770904-c712e886-1457-4770-9a46-f1bd83d180a3.png">
</p>

Closes #32682
Closes #32691
Closes #32694

This PR is made up of the following PRs:

- https://github.com/woocommerce/woocommerce/pull/32796
- https://github.com/woocommerce/woocommerce/pull/32852
- https://github.com/woocommerce/woocommerce/pull/32875
- https://github.com/woocommerce/woocommerce/pull/32913
- https://github.com/woocommerce/woocommerce/pull/32867
- https://github.com/woocommerce/woocommerce/pull/32935
- https://github.com/woocommerce/woocommerce/pull/32958

### How to test the changes in this Pull Request:

1. Checkout this branch and build assets by running `pnpm nx watch woocommerce-admin` _**or**_ grab a zip of this PR here → [woocommerce.zip](https://github.com/woocommerce/woocommerce/files/8705383/woocommerce.zip)
2. **Plugin eligibility**: Ensure the following plugins are not installed at all: 
    - WooCommerce Subscriptions
    - WooCommerce Payments
    - Sumo Subscriptions
    - Yith Subscriptions
    - Subscripto
    - Subscriptions For WooCommerce
3. **Store tracking eligibility:** Your store allows tracking.
    - This checkbox should be checked: **WooCommerce → Settings → Advanced → WooCommerce.com → Enable tracking**.
3. **Store location eligibility:** For the menu item to be displayed your store needs to based in the US. 
     - Go to **WooCommerce > Settings** and make sure your store is set to a US based address. 
3. **Store age criteria:** To view the subscription menu item your store will need to have been active for more than 6 months - we're testing this experiment on established stores. If your store doesn't meet this criteria, you can use the following code snippet: 
    - `update_option( 'woocommerce_admin_install_timestamp', strtotime( '-6 month' ) );`
    - **_or_** change the `woocommerce_admin_install_timestamp` option to `1620866357` (timestamp from ~1 year ago) via the `/wp-admin/options.php` page.
5. **Paid order criteria:** Only stores with at least 1 order in the last 30 days are eligible. If your store doesn't have an order in the last 30 days. Either: 
     1. Set any existing pending or failed order to processing or completed 
     2. If you don't have any order at all, you can just manually create one in **WooCommerce > Orders > Add new** and set it to processing or completed status
     3. ⚠️ **Note:** this order sale query is cached and so you may need to delete the `'woocommerce-wcpay-subscriptions_recent_sales_eligibility'` transient if this criteria failed previously. 
6. If you meet all the criteria above you should see a **WooCommerce > Subscriptions (new)** menu item.
7. View the **WooCommerce > Subscriptions (new)** page. 
     - The 'new' badge should disappear on subsequent page loads.

<p align="center">
<img width="322" alt="Screen Shot 2022-05-11 at 2 34 56 pm" src="https://user-images.githubusercontent.com/8490476/167772790-ef299bac-7d6b-4245-a640-5022c8c23e35.png">
</p>

#### The error scenario

If an error occurs during plugin installation and activation an error message will be displayed directing the user to install the WooCommerce Payments plugin manually. 

To test this flow:

1. Add the following code snippet to your site. 
2. Go to **WooCommerce -> Subscriptions**
4. Click **"Get started"**
5. An error should occur installing the plugin. See example below. 

```
define( 'DISALLOW_FILE_MODS', true );
```

<p align="center">
<img width="1200" alt="Screen Shot 2022-05-11 at 3 00 59 pm" src="https://user-images.githubusercontent.com/8490476/167772680-233c53f2-5871-4f29-b7e3-3a66d5a5ca75.png">
</p>

❗**Note:** The link takes the user to the [WooCommerce Payments wp.org page](https://wordpress.org/plugins/woocommerce-payments/).

#### The success scenario

1. Go to **WooCommerce > Subscriptions**
2. Click **"Get started"**
3. WooCommerce Payments will be installed and activated in the background
4. Once complete, you will be directed to the “Add new product” page or to the WCPay onboarding flow depending on which experiment treatment you were assigned to. Please see the "Testing the treatments" section below for more instructions.

#### The dismiss scenario

1. Go to **WooCommerce > Subscriptions**
2. Click **"No thanks"**
3. You should be redirected to the **WooCommerce > Home** screen and the subscriptions menu item should no longer be displayed.

**Note:** If you need to bring the menu item back, delete or change the `woocommerce-wcpay-subscriptions_dismissed` wp option.

#### Testing the treatments

Download, install and activate the [WooCommerce Admin Test Helper](https://github.com/woocommerce/woocommerce-admin-test-helper) plugin.

Navigate to **Tools → WCA Test Helper → Experiments**. This page should contain a row for the "woocommerce_wcpay_subscriptions_page_202207_v1" experiment.

Here you can toggle between the control (treatment A) and the treatment (treatment B).

#### Treatment A

1. Toggle to the control experience from **Tools → WCA Test Helper → Experiments**.
2. Navigate to **WooCommerce → Subscriptions**.
3. Observe the following step headers in the "You’re only steps away from selling subscriptions" section.
     1. Create a subscription
     2. Create and connect your account
     3. Provide a few business details
4. Click the "Get started" button.
5. The WooCommerce Payments plugin will be installed & activated and you will be redirected to the “Add new product” page.

#### Treatment B

1. Toggle to the treatment experience from **Tools → WCA Test Helper → Experiments**.
2. Navigate to **WooCommerce → Subscriptions**.
3. Observe the following step headers in the "You’re only steps away from selling subscriptions" section.
     1. Create and connect your account
     2. Provide a few business details
     3. Create subscriptions
4. Click the "Get started" button.
5. The WooCommerce Payments plugin will be installed & activated and you will be redirected to the WCPay onboarding flow.

--

<br class="Apple-interchange-newline">

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm nx changelog <project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Commit:
Add plumbing to render the "Subscription" admin menu item and page.

Commit:
Added placeholder content.

Commit:
Install WCPay on button click.

Commit:
Add tracking on button click.

james-allan committed:
Hide the notifications badge when admin have viewed the WCPay Subscriptions page

james-allan committed:
Reorder the WC menu items to place subscriptions after orders

Commit:
Merge branch 'add/experiment-subscriptions-admin-menu-2' into add/experiment-subscriptions-admin-menu-1

Commit:
Enqueue an inline script.

james-allan committed:
Display an error message when WC Payments fails to install

Jinksi committed:
Add method to check WcPaySubscriptionsPage eligibility

james-allan committed:
Direct the user to the wp.org wcpayments page for a more direct install path

james-allan committed:
Reword the error message to be more action -> outcome structure

Jinksi committed:
Add store activation date eligibility criteria

james-allan committed:
Replace translation domains from woocommerce-payments to woocommerce

james-allan committed:
Merge pull request #32852 from woocommerce/add/experiment-subscriptions-error-notice

Display an error message when WC Payments fails to install via the Subscriptions menu item

Jinksi committed:
Fix recent sales query, restrict to completed orders

Jinksi committed:
Fix store base location eligibility check to limit to USA

Jinksi committed:
Change plugin eligibility to check for installed plugins

Jinksi committed:
Reduce recent sales query scope to improve performance

Commit:
Implement the new subscriptions page design

Commit:
Merge branch 'add/experiment-subscriptions-admin-menu' into add/experiment-subscriptions-admin-menu-1

Commit:
Merge branch 'add/experiment-subscriptions-admin-menu-1' into add/experiment-subscriptions-admin-new-design

Jinksi committed:
Add sumosubscriptions to plugin eligibility check

Jinksi committed:
Add subscriptio to plugin eligibility check

Jinksi committed:
Add xa-woocommerce-subscriptions to plugin eligibility check

Co-authored-by: James Allan <james.allan@automattic.com>

Commit:
Fix typo.

Commit:
Fix font weight.

Jinksi committed:
Fix recent sales query to allow orders without completed status

Commit:
Merge pull request #32875 from woocommerce/add/experiment-subscriptions-admin-new-design

Implement the new subscriptions page design

Jinksi committed:
Cache recent sales eligibility to reduce excessive DB queries

james-allan committed:
Add a dismiss ("No thanks") button to the wcpay susbcriptions welcome page

james-allan committed:
Reorder and self close the No thanks buttons

james-allan committed:
Use updateOptions from @woocommerce/data OPTIONS_STORE_NAME to set wp options instead of using apiFetch

james-allan committed:
Merge pull request #32913 from woocommerce/add/experiment-subscriptions-admin-dismiss-button

Add a dismiss ("No thanks") button to the WCPay Subscriptions welcome page

Commit:
Change the badge wording to "new".

Commit:
Fix enqueue_scripts logic.

Jinksi committed:
Store recent sales eligibility transient as string

Jinksi committed:
Change recent sales eligibility transient string to yes/no

Commit:
Merge branch 'add/experiment-subscriptions-admin-menu-1' into add/experiment-subscriptions-admin-menu-eligibility

# Conflicts:
#	plugins/woocommerce/src/Internal/Admin/WcPaySubscriptionsPage.php

Jinksi committed:
Fix store sale eligibility comments to correct intent

Jinksi committed:
Use wc formatting functions for transient booleans

Commit:
Add @ts-expect-error comment to avoid build error.

Commit:
Add the wccore_subscriptions_empty_state_view tracking event.

james-allan committed:
Redirect to the create subscription product step after installing WC Payments

This is a temporary measure to be enable testing the full flow whilst we
work on setting up the experiment.

james-allan committed:
Merge pull request #32935 from woocommerce/add/experiment-subscriptions-redirect-after-install

Redirect to the create subscription product step after installing WC Payments via Subscriptions experiment page

Commit:
Add the wccore_subscriptions_empty_state_get_started_error tracking event.

Jinksi committed:
Merge pull request #32867 from woocommerce/add/experiment-subscriptions-admin-menu-eligibility

Prevent rendering the WooCommerce > Subscriptions menu item for WooCommerce stores that are ineligible for the experiment.

james-allan committed:
Remove unused maestro card SVG file

james-allan committed:
Add changelog

Commit:
Add basic experiment functionality.

james-allan commented:
If you would like to test these changes and just want a zip for a JN site or something else, here's one you can use: [woocommerce.zip](https://github.com/woocommerce/woocommerce/files/8667012/woocommerce.zip)

Commit:
Merge branch 'trunk' into add/experiment-subscriptions-admin-menu-1

Commit:
Tidy up experiment code and conditional rendering for the onboarding steps component.

Commit:
Hardcode experiment name.

Commit:
Copy changes and remove console log

Commit:
Merge pull request #32958 from woocommerce/add/experiment-subscriptions-admin-menu-experiment-framework

Add experiment functionality to the WooCommerce → Subscriptions page

botwoo commented:
:bar_chart: Test reports for 1fa306069703bb2d9c623d90c79835360039a059 `Merge pull request #32958 from woocommerce/add/experiment-subscriptio…` have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/32957/api/)
- E2E Test Report (TBC)

Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them.

Commit:
Merge branch 'trunk' into add/experiment-subscriptions-admin-menu-1

botwoo commented:
:bar_chart: Test reports for 0afc9dd2de2a903d8ff326f545a79dbfb5d16c2d `Merge branch 'trunk' into add/experiment-subscriptions-admin-menu-1` have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/32957/api/)
- E2E Test Report (TBC)

Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them.

joshuatf commented on plugins/woocommerce-admin/client/subscriptions/index.tsx:
```diff
@@ -0,0 +1,344 @@
+			onClick={ () => {
+				setIsNoThanksClicked( true );
+				recordEvent(
+					'wccore_subscriptions_empty_state_no_thanks_click',
```
Not sure if this was intentional, but just a heads up that these events are all prefixed with `wcadmin_` so the final event name will be `wcadmin_wccore_subscriptions_empty_state_no_thanks_click`.

Jinksi replied:
Thanks for the heads up @joshuatf, this is intentional. We're tracking the `wcadmin_wccore_subscriptions_empty_state_no_thanks_click` event with the `wcadmin_` prefix.

joshuatf commented on plugins/woocommerce-admin/client/subscriptions/index.tsx:
```diff
@@ -0,0 +1,344 @@
+
+const NoThanksButton = () => {
+	const [ isNoThanksClicked, setIsNoThanksClicked ] = useState( false );
+	const { updateOptions } = useDispatch( OPTIONS_STORE_NAME );
```
Just want to call out that we will hopefully deprecate the options API and datastores in the future.

However, we don't have a replacement API in place yet, so no changes necessary here right now.

joshuatf commented on plugins/woocommerce-admin/client/subscriptions/index.tsx:
```diff
@@ -0,0 +1,344 @@
+				'woocommerce'
+			) }
+		</h2>
+		<div className="subscriptions-page-onboarding-steps">
```
I don't know if we have a strict standard around this, but most of our components follow a BEM pattern for classes.

`subscriptions-page__onboarding-steps`

james-allan replied:
Thanks @joshuatf. I've fixed this and the other class names in [9166928](https://github.com/woocommerce/woocommerce/pull/32957/commits/91669281ec5aeee6d26019be20c7b413ccd8e2ee)

joshuatf commented on plugins/woocommerce-admin/client/layout/controller.js:
```diff
@@ -215,6 +221,22 @@ export const getPages = () => {
 		} );
 	}
 
+	if ( window.wcAdminFeatures[ 'wc-pay-subscriptions-page' ] ) {
```
Is the plan to take over this page in WC Pay once its activated?  I get that this page needs to be registered here since WC Pay is not yet installed at this point, but hoping post-install code can be extended within the WC Pay plugin.

aprea replied:
>Is the plan to take over this page in WC Pay once its activated?

That's correct. WooCommerce Payments registers its own "Subscriptions" admin menu/page via its bundled subscriptions functionality (i.e. https://github.com/automattic/woocommerce-subscriptions-core).

The Subscriptions page looks like this when WCPay is active:

<img width="1157" alt="Screen Shot 2022-05-18 at 9 57 00 am" src="https://user-images.githubusercontent.com/1527164/168930880-5e4c05b3-34f2-4ca5-8380-7321c4c862e6.png">

>but hoping post-install code can be extended within the WC Pay plugin.

Could you clarify this point? I'm not sure what you mean.

joshuatf replied:
>Could you clarify this point? I'm not sure what you mean.

Meaning that the actual Subscriptions page is housed in WC Pay after install and overrides the one registered here, which looks to be the case from your above comment. 😄

Commit:
Only register the the subscriptions page if tracking is allowed

james-allan committed:
Rename component class names with BEM standard names

james-allan committed:
Fix typo

vedanshujain commented on plugins/woocommerce/src/Internal/Admin/WcPaySubscriptionsPage.php:
```diff
@@ -0,0 +1,266 @@
+		add_action( 'admin_enqueue_scripts', array( $this, 'enqueue_scripts' ) );
+
+		// Priority 50 to run after Automattic\WooCommerce\Internal\Admin\Homescreen::update_link_structure() which runs on 20.
+		add_action( 'admin_menu', array( $this, 'restructure_menu_order' ), 50 );
```
if this should always be called after `Admin\Homescreen::update_link_structure`, it might be better to call this directly. There could be other menu items interfering between priority `20` from `update_link_structure` and `50` here.

james-allan replied:
Just to clarify, you recommend we call our function to move the subscriptions menu item (`WcPaySubscriptionsPage::restructure_menu_order()`) to be called directly from within the (`Admin\Homescreen::update_link_structure()`) function?

We can definitely do that however the `Admin\Homescreen` class specifically handles the "Home" menu item and so it would be additional code clutter for it to also be calling a function directly to wrangle another menu item. 

> There could be other menu items interfering between priority 20 from `update_link_structure` and 50 here.

I don't think this is a big concern for us if this were to happen. We'd prefer that this "Subscriptions" menu item has consistant placement with the WooCommerce Subscriptions "subscriptions" menu item, however, if something was to interfere and move it further down or up we wouldn't be too concerned. I'd be more concerned about a less cleaner code separation tbh given this is an experiment.

vedanshujain commented on plugins/woocommerce/src/Internal/Admin/WcPaySubscriptionsPage.php:
```diff
@@ -0,0 +1,266 @@
+		// Get a single order that has been paid within the last 30 days.
+		$orders = wc_get_orders(
+			array(
+				'date_paid' => '>' . strtotime( '-30 days' ),
```
let's use `date_created` instead of paid since `date_paid` is stored in the metadata, so this query will be slower. you can add another filter on status for `processed` and `completed` if you'd like to get only orders that are paid.

james-allan replied:
👍 can do.

I've changed to use the `'date_created'` and `'status'` args instead (https://github.com/woocommerce/woocommerce/pull/32957/commits/3aa2294968cfdf8d1896681210fa2daa0cfb8e16)

vedanshujain commented on plugins/woocommerce/src/Internal/Admin/WcPaySubscriptionsPage.php:
```diff
@@ -0,0 +1,266 @@
+			'title'      => _x( 'Subscriptions', 'Admin menu name', 'woocommerce' ),
+			'parent'     => 'woocommerce',
+			'path'       => '/subscriptions',
+			'capability' => 'manage_options',
```
should this be `manage_woocommerce`?

aprea replied:
The new "Subscriptions" admin page has a "Get started" button that, when clicked, will install and activate the WooCommerce Payments plugin on the site. The user must then complete WCPay onboarding (Stripe KYC).

So this capability needs to be permissive enough to allow them to complete these actions.

I just checked the `manage_woocommerce` capability and I don't think it'll work here as it's granted to the [Shop Manager](https://woocommerce.com/document/roles-capabilities/) role. This role **does not** have the required `activate_plugins` or `install_plugins` capabilities. ([ref](https://github.com/woocommerce/woocommerce/blob/ee01d4219282387c2975ef4594677453c1dd7a0e/includes/class-wc-install.php#L1052-L1097))

vedanshujain commented on plugins/woocommerce/src/Internal/Admin/WcPaySubscriptionsPage.php:
```diff
@@ -0,0 +1,266 @@
+			}
+
+			// Keep a record of the subscriptions item and remove it from its current place in the menu.
+			if ( 'wc-admin&path=/subscriptions' === $menu_item[2] ) {
```
this string `wc-admin&path=/subscriptions` seems to be used in multiple places, should we define a constant for this?

james-allan replied:
👍 Fixed in [3cecb3d](https://github.com/woocommerce/woocommerce/pull/32957/commits/3cecb3d72c8a4b84f5b0990e9a2a9d7c2dc7275e)

vedanshujain commented on plugins/woocommerce/src/Internal/Admin/WcPaySubscriptionsPage.php:
```diff
@@ -0,0 +1,266 @@
+	 */
+	private function is_store_experiment_eligible() {
+		// Ineligible if WooCommerce Payments OR an existing subscriptions plugin is installed.
+		$installed_plugins      = PluginsHelper::get_installed_plugin_slugs();
```
is the `installed_plugins` here intentional? or should this be activated plugins instead?

Jinksi replied:
We're filtering out stores who have any of these plugins installed whether activated or not, so `installed_plugins` is intentional here 👍

vedanshujain commented on plugins/woocommerce/src/Internal/Admin/WcPaySubscriptionsPage.php:
```diff
@@ -0,0 +1,266 @@
+
+		$experiment_assignment = 'A';
+
+		if ( $this->is_store_experiment_eligible() ) {
```
shouldn't we bail here from enqueuing the script if the store is not eligible? or do we need this script either way?

aprea replied:
Good catch, thanks! Fixed in 98e23a8

vedanshujain reviewed (commented):
Testing well, added a few questions, though. Thanks!

Commit:
Only add the inline script if the store is eligible for the experiment

james-allan committed:
Use date created and status order filter for order eligiblity check

Using these two order properties instead of the paid_date field is more performance friendly
as both fields are in the wp-posts table - not in the post_meta table.

ref: https://github.com/woocommerce/woocommerce/pull/32957/files/71e9657e52d7d6c8bee863ca064bb2153ee11798#r875545984

james-allan committed:
Introduce susbcriptions menu item slug constant

aprea commented:
Hi @vedanshujain, apologies for the late notice but we were hoping to have this PR merged in time for the WC 6.6. code freeze (Thursday 4pm UTC).

Would you be able to approve/merge this PR assuming there's no additional blocking feedback?

/cc @haszari
